Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m

```diff
@@ -469,26 +469,23 @@ - (instancetype)initWithLink:(NSURL *)link domain:(NSString *)domain {
   }
   NSString *domainURIPrefix =
       domainURL.scheme ? domain : [NSString stringWithFormat:@"https://%@", domain];
-  return [self initWithLink:link domainURIPrefix:domainURIPrefix];
+  return [self initWithLink:link domainURIPrefix:domainURIPrefix allowNilReturn:NO];
 }
 
 #pragma mark Initializers.
 + (instancetype)componentsWithLink:(NSURL *)link domainURIPrefix:(NSString *)domainURIPrefix {
-  NSURL *domainURIPrefixURL = [NSURL URLWithString:domainURIPrefix];
-  if (!domainURIPrefixURL) {
-    FDLLog(FDLLogLevelError, FDLLogIdentifierSetupInvalidDomainURIPrefix,
-           @"Invalid domainURIPrefix. Please input a valid URL.");
-    return nil;
-  }
-  if (![[domainURIPrefixURL.scheme lowercaseString] isEqualToString:@"https"]) {
-    FDLLog(FDLLogLevelError, FDLLogIdentifierSetupInvalidDomainURIPrefixScheme,
-           @"Invalid domainURIPrefix scheme. Scheme needs to be https");
-    return nil;
-  }
   return [[self alloc] initWithLink:link domainURIPrefix:domainURIPrefix];
 }
 
 - (instancetype)initWithLink:(NSURL *)link domainURIPrefix:(NSString *)domainURIPrefix {
+  return [self initWithLink:link domainURIPrefix:domainURIPrefix allowNilReturn:YES];
+}
+
+/// Deprecated init API has nonnull return type. If the check fails, support returning an instance
+/// for the deprecated API (the backend call will return an error. This is pre-existing behavior).
+- (instancetype)initWithLink:(NSURL *)link
+             domainURIPrefix:(NSString *)domainURIPrefix
+              allowNilReturn:(BOOL)allowNilReturn {
```

Comment:
This really only assigns two private variables but does validation - I think it's worth just keeping the old constructor as-is (before this change) and assign the two internal variables without validation, and leave the logic inside the new one to return `nil` if it doesn't validate.

Reply:
It also does [super init]. With my change, the initWithLink:domainURIPrefix:allowNilReturn: becomes the designated initializer if you will.

Reply:
Updated to offline conv: keep deprecated initializer unchanged.